### PR TITLE
[GISEL] Add IRTranslation for shufflevector on scalable vector types

### PR DESCRIPTION
**Pull request by michaelmaitland**
This patch is stacked on https://github.com/llvm/llvm-project/pull/80372, https://github.com/llvm/llvm-project/pull/80307, and https://github.com/llvm/llvm-project/pull/80306.

ShuffleVector on scalable vector types gets IRTranslate'd to G_SPLAT_VECTOR since a ShuffleVector that has operates on scalable vectors is a splat vector where the value of the splat vector is the 0th element of the first operand, since the index mask operand is the zeroinitializer (undef and poison are treated as zeroinitializer here). This is analogous to what happens in SelectionDAG for ShuffleVector.

`buildSplatVector` is renamed to`buildBuildVectorSplatVector`. I did not make this a separate patch because it would cause problems to revert that change without reverting this change too.

### CODE REVIEW

File: llvm/lib/CodeGen/MachineVerifier.cpp [hunk 1]
```diff
@@ -1618,20 +1618,34 @@ void MachineVerifier::verifyPreISelGenericInstruction(const MachineInstr *MI) {
...
+               MI);
+    } else {
+      // Idxes for fixed vectors must be in bounds or undef, which is
+      // represented as -1.
+      for (int Idx : MaskIdxes) {
+        if (Idx < 0)
+          continue;
+        if ((unsigned)Idx >= 2 * SrcNumElts.getFixedValue())
+          report("Out of bounds shuffle index", MI);
+      }
```
Comment (arsenm) [inline llvm/lib/CodeGen/MachineVerifier.cpp:1648]: Should add some tests in test/MachineVerifier for this

Comment (michaelmaitland) [reply llvm/lib/CodeGen/MachineVerifier.cpp]: Added

**Comment by topperc** [comment]
SelectionDAG doesn't support ISD::SHUFFLEVECTOR for scalable vectors. It only supports ISD::SPLAT_VECTOR. Are we trying to do something different for GISel for some reason?

**Comment by michaelmaitland** [comment]
> SelectionDAG doesn't support ISD::SHUFFLEVECTOR for scalable vectors. It only supports ISD::SPLAT_VECTOR. Are we trying to do something different for GISel for some reason?

I intend to write a patch after this that combines g_insert_element + g_shuffle_vector into g_splat_vector where appropriate. For scalable vectors, that is in all cases since the mask index vector is always zeroinit (undef case can be treated as zeroinit). 

I thought there is a case where fixed vectors can generate a non-splat vector which is why we kept the opcode. But I will defer to someone who knows more about this to answer why it’s currently different than SDAG.

**Comment by topperc** [comment]
> > SelectionDAG doesn't support ISD::SHUFFLEVECTOR for scalable vectors. It only supports ISD::SPLAT_VECTOR. Are we trying to do something different for GISel for some reason?
> 
> 
> 
> I intend to write a patch after this that combines g_insert_element + g_shuffle_vector into g_splat_vector where appropriate. For scalable vectors, that is in all cases since the mask index vector is always zeroinit (undef case can be treated as zeroinit). 
> 
> 

What if the G_INSERT_ELT doesn't exist? SelectionDAG always creates an extract+splat_vector when it translates a scalable shufflevector.

**Comment by michaelmaitland** [comment]
> > > SelectionDAG doesn't support ISD::SHUFFLEVECTOR for scalable vectors. It only supports ISD::SPLAT_VECTOR. Are we trying to do something different for GISel for some reason?
> > 
> > 
> > I intend to write a patch after this that combines g_insert_element + g_shuffle_vector into g_splat_vector where appropriate. For scalable vectors, that is in all cases since the mask index vector is always zeroinit (undef case can be treated as zeroinit).
> 
> What if the G_INSERT_ELT doesn't exist? SelectionDAG always creates an extract+splat_vector when it translates a scalable shufflevector.

Now that I think more about it, I am pretty sure that G_SHUFFLE_VECTOR for scalable vectors is the same thing as a splat, since the mask is always taking from index zero. If there is no insertelement before it, then we are splatting whatever is in index 0 of the input vector.

**Comment by topperc** [comment]
> > > > SelectionDAG doesn't support ISD::SHUFFLEVECTOR for scalable vectors. It only supports ISD::SPLAT_VECTOR. Are we trying to do something different for GISel for some reason?
> > > 
> > > 
> > > I intend to write a patch after this that combines g_insert_element + g_shuffle_vector into g_splat_vector where appropriate. For scalable vectors, that is in all cases since the mask index vector is always zeroinit (undef case can be treated as zeroinit).
> > 
> > 
> > What if the G_INSERT_ELT doesn't exist? SelectionDAG always creates an extract+splat_vector when it translates a scalable shufflevector.
> 
> Now that I think more about it, I am pretty sure that G_SHUFFLE_VECTOR for scalable vectors is the same thing as a splat, since the mask is always taking from index zero. If there is no insertelement before it, then we are splatting whatever is in index 0 of the input vector.

Yes the only scalable shufflevector supported is a splat of element 0. The different between ISD::SHUFFLE_VECTOR and ISD::SPLAT_VECTOR is that ISD::SHUFFLE_VECTOR takes a vector input and ISD::SPLAT_VECTOR takes a scalar input.

**Comment by michaelmaitland** [comment]
> Yes the only scalable shufflevector supported is a splat of element 0. The difference between ISD::SHUFFLE_VECTOR and ISD::SPLAT_VECTOR is that ISD::SHUFFLE_VECTOR takes a vector input and ISD::SPLAT_VECTOR takes a scalar input.

Do you think it would be reasonable to deal with converting a G_SHUFFLE_VECTOR into a G_SPLAT_VECTOR during legalization? Or are you asking for it to occur here during IRTranslation? FYI there is no G_SPLAT_VECTOR opcode today -- we would need to add that.

**Comment by tschuett** [comment]
GlobalIsel has no splat_vector yet. I was hoping that https://github.com/llvm/llvm-project/pull/74502 would make progress.

**Comment by michaelmaitland** [comment]
> GlobalIsel has no splat_vector yet. I was hoping that #74502 would make progress.

We might still want a G_SPLAT_VECTOR in the case that the splat is not constant. WDYT?

**Comment by tschuett** [comment]
@aemerson We would still need G_FSPLAT_VECTOR and G_SPLAT_VECTOR. If there are constants or operations?!?

**Comment by tschuett** [comment]
IR gets splats as constants. The IR translator eventually has to translate them into GlobalIsel.

**Comment by tschuett** [comment]
How would you otherwise distinguish between integer and float splats? If a splat vector only gets an immediate?

**Comment by topperc** [comment]
> How would you otherwise distinguish between integer and float splats? If a splat vector only gets an immediate?

How do we do it for a BUILD_VECTOR of immediates?

**Comment by arsenm** [comment]
> How would you otherwise distinguish between integer and float splats? If a splat vector only gets an immediate?

There's no need to, but you can always inspect the operands. In fact, I think we should get rid of G_FCONSTANT altogether and just adds unnecessary hassle

**Comment by tschuett** [comment]
```
   %v8:_(<2 x s64>) = G_BUILD_VECTOR %0(s64), %1(s64)
```
A build vector could be mixed between integer and float.

File: llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h [hunk 2]
```diff
@@ -1063,8 +1063,8 @@ class MachineIRBuilder {
 
   /// Build and insert \p Res = G_BUILD_VECTOR with \p Src replicated to fill
   /// the number of elements
-  MachineInstrBuilder buildSplatVector(const DstOp &Res,
-                                       const SrcOp &Src);
+  MachineInstrBuilder buildBuildVectorSplatVector(const DstOp &Res,
```
Comment (topperc) [inline llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h:1066]: Maybe `buildSplatBuildVector`. Similar to SelectionDAG's `getSplatBuildVector`

Comment (tschuett) [reply llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h:1066]: `buildConstant` can build scalars and splat vectors. It depends on the `type` that you feed in.

**Comment by tschuett** [comment]
Could you please add the new opcode to https://github.com/llvm/llvm-project/blob/main/llvm/docs/GlobalISel/GenericOpcode.rst? Thanks.

File: llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h (hunk 2 above)
Comment (michaelmaitland) [reply llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h:1066]: Updated to buildSplatBuildVector.

@tschuett I am not sure where buildConstant fits in here. Could you clarify?

**Comment by michaelmaitland** [comment]
> Could you please add the new opcode to https://github.com/llvm/llvm-project/blob/main/llvm/docs/GlobalISel/GenericOpcode.rst? Thanks.

Done

File: llvm/docs/GlobalISel/GenericOpcode.rst [hunk 3]
```diff
@@ -639,6 +639,11 @@ Concatenate two vectors and shuffle the elements according to the mask operand.
 The mask operand should be an IR Constant which exactly matches the
 corresponding mask for the IR shufflevector instruction.
 
+G_SPLAT_VECTOR
+^^^^^^^^^^^^^^^^
+
+Create a vector where all elements are the scalar from the source operand.
+
```
Comment (arsenm) [inline llvm/docs/GlobalISel/GenericOpcode.rst:646]: Unless we do the same in the DAG, I think this just adds complication down the line for pattern sharing. The buildBuildSplatVector can just create the G_BUILD_VECTOR with N copies of the input

**Review by arsenm — APPROVED** [review]

**Comment by adrian-prantl** [comment]
@michaelmaitland I'm getting crashes of tblgen when building ToT main and reverting this patch seems to fix it. Could you please revert this once more and investigate?

**Pull request merged**